# Hacktiv8 Phase 1: Graded Challenge 1

---

Graded Challenge ini dibuat guna mengevaluasi pembelajaran pada Hacktiv8 Data Science Fulltime Program khususnya pada konsep Regression.

## Introduction

By [Rifky Aliffa](https://github.com/Penzragon)

![Uber vs Lyft](https://photos5.appleinsider.com/gallery/37267-69909-1OLXiBSIR4zfrwXBI_emrgQ-xl.jpg)

### Dataset

Dataset yang digunakan pada project ini adalah dataset yang berisi tentang data perjalanan penumpang menggunakan **Uber** dan **Lyft** yang sudah termasuk informasi cuaca pada jam tersebut. Dataset ini berisi 693071 baris dengan 57 kolom yang di antaranya adalah id, timestamp, hour, day, month, datetime, timezone, source, destination, cab_type, product_id, name, price, distance, dan masih banyak lagi. Dataset dapat dilihat di [Kaggle](https://www.kaggle.com/brllrb/uber-and-lyft-dataset-boston-ma).

### Objectives

**Graded Challenge 1** ini dibuat guna mengevaluasi konsep Regression sebagai berikut:

- Mampu memahami konsep regression dengan Linear Regession
- Mampu mempersiapkan data untuk digunakan dalam model Linear Regression
- Mampu mengimplementasikan Linear Regression untuk membuat prediksi

## Import Libraries

Pada project kali ini library yang akan digunakan adalah **Pandas**, **Numpy**, **Matplotlib**, **Seaborn**, dan **Scikit-Learn**.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

Melakukan setting agar pandas menampilkan seluruh kolom pada dataframe.

In [2]:
pd.set_option('display.max_columns', None)

## Data Loading

membuat dataframe dari file `rideshare_kaggle.csv`.

In [3]:
rides = pd.read_csv('rideshare_kaggle.csv')

In [4]:
rides.head()

id     timestamp  hour  day  month  \
0  424553bb-7174-41ea-aeb4-fe06d4f4b9d7  1.544953e+09     9   16     12   
1  4bd23055-6827-41c6-b23b-3c491f24e74d  1.543284e+09     2   27     11   
2  981a3613-77af-4620-a42a-0c0866077d1e  1.543367e+09     1   28     11   
3  c2d88af2-d278-4bfd-a8d0-29ca77cc5512  1.543554e+09     4   30     11   
4  e0126e1f-8ca9-4f2e-82b3-50505a09db9a  1.543463e+09     3   29     11   

              datetime          timezone            source    destination  \
0  2018-12-16 09:30:07  America/New_York  Haymarket Square  North Station   
1  2018-11-27 02:00:23  America/New_York  Haymarket Square  North Station   
2  2018-11-28 01:00:22  America/New_York  Haymarket Square  North Station   
3  2018-11-30 04:53:02  America/New_York  Haymarket Square  North Station   
4  2018-11-29 03:49:20  America/New_York  Haymarket Square  North Station   

  cab_type    product_id          name  price  distance  surge_multiplier  \
0     Lyft     lyft_line        Shared    5.0      0.44               1.0   
1     Lyft  lyft_premier           Lux   11.0      0.44               1.0   
2     Lyft          lyft          Lyft    7.0      0.44               1.0   
3     Lyft   lyft_luxsuv  Lux Black XL   26.0      0.44               1.0   
4     Lyft     lyft_plus       Lyft XL    9.0      0.44               1.0   

   latitude  longitude  temperature  apparentTemperature    short_summary  \
0   42.2148    -71.033        42.34                37.12   Mostly Cloudy    
1   42.2148    -71.033        43.58                37.35            Rain    
2   42.2148    -71.033        38.33                32.93           Clear    
3   42.2148    -71.033        34.38                29.63           Clear    
4   42.2148    -71.033        37.44                30.88   Partly Cloudy    

                                        long_summary  precipIntensity  \
0                          Rain throughout the day.            0.0000   
1   Rain until morning, starting again in the eve...           0.1299   
2                        Light rain in the morning.            0.0000   
3                 Partly cloudy throughout the day.            0.0000   
4                 Mostly cloudy throughout the day.            0.0000   

   precipProbability  humidity  windSpeed  windGust  windGustTime  visibility  \
0                0.0      0.68       8.66      9.17    1545015600      10.000   
1                1.0      0.94      11.98     11.98    1543291200       4.786   
2                0.0      0.75       7.33      7.33    1543334400      10.000   
3                0.0      0.73       5.28      5.28    1543514400      10.000   
4                0.0      0.70       9.14      9.14    1543446000      10.000   

   temperatureHigh  temperatureHighTime  temperatureLow  temperatureLowTime  \
0            43.68           1544968800           34.19          1545048000   
1            47.30           1543251600           42.10          1543298400   
2            47.55           1543320000           33.10          1543402800   
3            45.03           1543510800           28.90          1543579200   
4            42.18           1543420800           36.71          1543478400   

   apparentTemperatureHigh  apparentTemperatureHighTime  \
0                    37.95                   1544968800   
1                    43.92                   1543251600   
2                    44.12                   1543320000   
3                    38.53                   1543510800   
4                    35.75                   1543420800   

   apparentTemperatureLow  apparentTemperatureLowTime                   icon  \
0                   27.39                  1545044400   partly-cloudy-night    
1                   36.20                  1543291200                  rain    
2                   29.11                  1543392000           clear-night    
3                   26.20                  1543575600           clear-night    
4                   30.29                

### Data Characteristics

In [5]:
rides.shape

(693071, 57)

Dataframe ini terbentuk dari:
- **693071** baris
- **57** kolom

In [6]:
rides.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 693071 entries, 0 to 693070
Data columns (total 57 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           693071 non-null  object 
 1   timestamp                    693071 non-null  float64
 2   hour                         693071 non-null  int64  
 3   day                          693071 non-null  int64  
 4   month                        693071 non-null  int64  
 5   datetime                     693071 non-null  object 
 6   timezone                     693071 non-null  object 
 7   source                       693071 non-null  object 
 8   destination                  693071 non-null  object 
 9   cab_type                     693071 non-null  object 
 10  product_id                   693071 non-null  object 
 11  name                         693071 non-null  object 
 12  price                        637976 non-null  float64
 13 

Dari basic info ini dapat diketahui dataframe ini terdiri dari:
- 29 kolom dengan tipe float
- 17 kolom dengan tipe integer
- 11 kolom dengan tipe object

Dapat diketahui juga terdapat **missing value** pada kolom `price`.

In [7]:
rides.describe().T

count          mean            std  \
timestamp                    693071.0  1.544046e+09  689192.492586   
hour                         693071.0  1.161914e+01       6.948114   
day                          693071.0  1.779436e+01       9.982286   
month                        693071.0  1.158668e+01       0.492429   
price                        637976.0  1.654513e+01       9.324359   
distance                     693071.0  2.189430e+00       1.138937   
surge_multiplier             693071.0  1.013870e+00       0.091641   
latitude                     693071.0  4.233817e+01       0.047840   
longitude                    693071.0 -7.106615e+01       0.020302   
temperature                  693071.0  3.958439e+01       6.726084   
apparentTemperature          693071.0  3.588122e+01       7.918707   
precipIntensity              693071.0  8.922153e-03       0.026901   
precipProbability            693071.0  1.460758e-01       0.328938   
humidity                     693071.0  7.411187e-01       0.138595   
windSpeed                    693071.0  6.186253e+00       3.148223   
windGust                     693071.0  8.469860e+00       5.289179   
windGustTime                 693071.0  1.544049e+09  692824.402315   
visibility                   693071.0  8.467970e+00       2.602888   
temperatureHigh              693071.0  4.504098e+01       5.996541   
temperatureHighTime          693071.0  1.544050e+09  693792.107470   
temperatureLow               693071.0  3.415074e+01       6.383163   
temperatureLowTime           693071.0  1.544102e+09  692292.331984   
apparentTemperatureHigh      693071.0  4.161405e+01       7.666138   
apparentTemperatureHighTime  693071.0  1.544050e+09  694169.873815   
apparentTemperatureLow       693071.0  3.013982e+01       8.057468   
apparentTemperatureLowTime   693071.0  1.544099e+09  692737.825155   
dewPoint                     693071.0  3.165626e+01       9.142355   
pressure                     693071.0  1.010095e+03      13.472899   
windBearing                  693071.0  2.200559e+02      99.102736   
cloudCover                   693071.0  6.865020e-01       0.358534   
uvIndex                      693071.0  2.489081e-01       0.474097   
visibility.1                 693071.0  8.467970e+00       2.602888   
ozone                        693071.0  3.135136e+02      27.953061   
sunriseTime                  693071.0  1.544027e+09  691139.271721   
sunsetTime                   693071.0  1.544060e+09  690663.392963   
moonPhase                    693071.0  5.790655e-01       0.244705   
precipIntensityMax           693071.0  3.737441e-02       0.055214   
uvIndexTime                  693071.0  1.544044e+09  691202.767272   
temperatureMin               693071.0  3.345777e+01       6.467224   
temperatureMinTime           693071.0  1.544042e+09  690195.441408   
temperatureMax               693071.0  4.526131e+01       5.645046   
temperatureMaxTime           693071.0  1.544047e+09  690135.337190   
apparentTemperatureMin       693071.0  2.973100e+01       7.110494   
apparentTemperatureMinTime   693071.0  1.544048e+09  687186.191494   
apparentTemperatureMax       693071.0  4.199734e+01       6.936841   
apparentTemperatureMaxTime   693071.0  1.544048e+09  691077.651975   

                                      min           25%           50%  \
timestamp                    1.543204e+09  1.543444e+09  1.543737e+09   
hour                         0.000000e+00  6.000000e+00  1.200000e+01   
day                          1.000000e+00  1.300000e+01  1.700000e+01   
month                        1.100000e+01  1.100000e+01  1.200000e+01   
price                        2.500000e+00  9.000000e+00  1.350000e+01   
distance                     2.000000e-02  1.280000e+00  2.160000e+00   
surge_multiplier             1.000000e+00  1.000000e+00  1.000000e+00   
latitude                     4.221480e+01  4.235030e+01  4.235190e+01   
longitude                   -7.110540e+01 -7.108100e+01 -7.106310e+01   
temperature

Distribusi dari kolom pada dataset kebanyakan memiliki distribusi yang cenderung normal (**Symmetrical Distribution**) contohnya pada kolom `hour`, `distance`, dan `temperature` dimana mean dari kolom tersebut berdekatan dengan mediannya.

## Data Cleansing

### Missing Value

Pada bagian ini akan dilakukan pengecekan apakah ada missing value pada setiap kolom dataframe.

In [8]:
rides.isna().sum()

id                                 0
timestamp                          0
hour                               0
day                                0
month                              0
datetime                           0
timezone                           0
source                             0
destination                        0
cab_type                           0
product_id                         0
name                               0
price                          55095
distance                           0
surge_multiplier                   0
latitude                           0
longitude                          0
temperature                        0
apparentTemperature                0
short_summary                      0
long_summary                       0
precipIntensity                    0
precipProbability                  0
humidity                           0
windSpeed                          0
windGust                           0
windGustTime                       0
v

In [9]:
print(f'Percentage of missing values in price column: {(rides.price.isna().sum()/len(rides)*100):.2f}%')

Percentage of missing values in price column: 7.95%


Terdapat **55095** missing value pada kolom `price` dan persentase dari missing value pada kolom ini sebesar **7.95%**. 

In [10]:
rides[rides.price.isna()].head()

id     timestamp  hour  day  month  \
18  fa5fb705-03a0-4eb9-82d9-7fe80872f754  1.543674e+09    14    1     12   
31  eee70d94-6706-4b95-a8ce-0e34f0fa8f37  1.543795e+09    23    2     12   
40  7f47ff53-7cf2-4a6a-8049-83c90e042593  1.543524e+09    20   29     11   
60  43abdbe4-ab9e-4f39-afdc-31cfa375dc25  1.544732e+09    20   13     12   
69  80db1c49-9d51-4575-a4f4-1ec23b4d3e31  1.543583e+09    13   30     11   

               datetime          timezone         source    destination  \
18  2018-12-01 14:13:04  America/New_York      North End       West End   
31  2018-12-02 23:52:56  America/New_York    Beacon Hill  South Station   
40  2018-11-29 20:38:05  America/New_York  North Station         Fenway   
60  2018-12-13 20:10:16  America/New_York      North End       West End   
69  2018-11-30 13:08:03  America/New_York      North End    Beacon Hill   

   cab_type                            product_id  name  price  distance  \
18     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a  Taxi    NaN      1.11   
31     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a  Taxi    NaN      2.48   
40     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a  Taxi    NaN      2.94   
60     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a  Taxi    NaN      1.16   
69     Uber  8cf7e821-f0d3-49c6-8eba-e679c0ebcf6a  Taxi    NaN      2.67   

    surge_multiplier  latitude  longitude  temperature  apparentTemperature  \
18               1.0   42.3647   -71.0542        36.99                32.27   
31               1.0   42.3588   -71.0707        48.83                48.83   
40               1.0   42.3661   -71.0631        42.50                36.47   
60               1.0   42.3647   -71.0542        33.15                30.60   
69               1.0   42.3647   -71.0542        31.80                31.80   

      short_summary                                long_summary  \
18   Partly Cloudy    Light rain in the morning and overnight.    
31        Overcast                   Light rain until evening.    
40   Partly Cloudy           Partly cloudy throughout the day.    
60   Mostly Cloudy           Partly cloudy throughout the day.    
69           Clear           Mostly cloudy throughout the day.    

    precipIntensity  precipProbability  humidity  windSpeed  windGust  \
18              0.0                0.0      0.68       5.87      6.26   
31              0.0                0.0      0.96       2.83      2.98   
40              0.0                0.0      0.54      10.68     17.15   
60              0.0                0.0      0.57       3.02      3.02   
69              0.0                0.0      0.83       2.10      4.79   

    windGustTime  visibility  temperatureHigh  temperatureHighTime  \
18    1543672800       9.910            44.66           1543690800   
31    1543755600       7.790            50.73           1543788000   
40    1543514400       9.772            44.66           1543510800   
60    1544738400       9.956            33.83           1544731200   
69    1543554000      10.000            42.52           1543600800   

    temperatureLow  temperatureLowTime  apparentTemperatureHigh  \
18           35.04          1543712400                    43.99   
31           44.99          1543816800                    50.06   
40           28.68          1543575600                    38.30   
60           27.27          1544781600                    32.85   
69           31.71          1543658400                    40.53   

    apparentTemperatureHighTime  apparentTemperatureLow  \
18                   1543690800                   35.69   
31                   1543788000                   45.64   
40                   1543510800                   27.26   
60                   1544734800                   24.61   
69                   1543611600                   28.06   

    apparentTemperatureLowTime                 icon  dewPoint  pressure  \
18                  1543712400   partly-cloudy-day      27.53   1022.32   
31                  1543816

In [11]:
rides[rides.price.isna()].name.value_counts()

Taxi    55095
Name: name, dtype: int64

In [12]:
rides[rides.price.isna()].cab_type.value_counts()

Uber    55095
Name: cab_type, dtype: int64

Pada dataset seluruh data yang memiliki **missing value** pada kolom `price` adalah data dari penumpang yang menggunakan **Taxi** dari **Uber**. Tetapi karena missing value ini berada pada kolom **target** (yang akan dijadikan prediksi), maka seluruh missing value akan dihapus dari dataset karena nantinya akan mengganggu akurasi dari model.

In [13]:
rides = rides.dropna()

In [14]:
rides.isna().any()

id                             False
timestamp                      False
hour                           False
day                            False
month                          False
datetime                       False
timezone                       False
source                         False
destination                    False
cab_type                       False
product_id                     False
name                           False
price                          False
distance                       False
surge_multiplier               False
latitude                       False
longitude                      False
temperature                    False
apparentTemperature            False
short_summary                  False
long_summary                   False
precipIntensity                False
precipProbability              False
humidity                       False
windSpeed                      False
windGust                       False
windGustTime                   False
v

Setelah dilakukan lakukan penghapusan missing value pada kolom `price`, sekarang sudah tidak lagi terdapat missing value.

### Duplicated Data

In [15]:
rides.duplicated().sum()

0

In [16]:
rides.duplicated().any()

False

Tidak terdapat duplicated rows pada dataset ini.

## Exploratory Data Analysis (EDA)